src/rgw/rgw_rest_pubsub.cc

```diff
@@ -0,0 +1,461 @@
+      return;
+    }
+    const auto topic_it = bucket_topics.topics.find(sub_conf.topic);
+    if (topic_it == bucket_topics.topics.end()) {
+      op_ret = -ENOENT;
+      ldout(s->cct, 1) << "notification does not have topic information, ret=" << op_ret << dendl;
+      return;
+    }
+    add_notification_to_list(sub_conf, topic_it->second.events, topic_it->second.topic.arn);
+    return;
+  }
+  // get info on all s3 notifications of the bucket
+  rgw_pubsub_bucket_topics bucket_topics;
+  b->get_topics(&bucket_topics);
+  // loop through all topics of the bucket
+  for (const auto& topic : bucket_topics.topics) {
+    // for each topic get all subscriptions
+    rgw_pubsub_topic_subs topic_subs;
+    ups->get_topic(topic.first, &topic_subs);
+    const auto& events = topic.second.events;
+    const auto& topic_arn = topic.second.topic.arn;
+    // loop through all subscriptions
+    for (const auto& topic_sub_name : topic_subs.subs) {
+      // get info of a specific notification
+      auto sub = ups->get_sub(topic_sub_name);
+      rgw_pubsub_sub_config sub_conf;
+      op_ret = sub->get_conf(&sub_conf);
+      if (op_ret < 0) {
+        ldout(s->cct, 1) << "failed to get notification info, ret=" << op_ret << dendl;
+        return;
+      }
+      if (sub_conf.s3_id.empty()) {
+        // not an s3 notification
+        continue;
+      }
+      add_notification_to_list(sub_conf, events, topic_arn);
+    }
+  }
+}
+
+// s3 compliant notification handler factory
+class RGWHandler_REST_PSNotifs_S3 : public RGWHandler_REST_S3 {
+protected:
+  int init_permissions(RGWOp* op) override {
+    return 0;
+  }
+
+  int read_permissions(RGWOp* op) override {
+    return 0;
+  }
+  bool supports_quota() override {
+    return false;
+  }
+  RGWOp *op_get() override {
+    return new RGWPSListNotifs_ObjStore_S3();
+  }
+  RGWOp *op_put() override {
+    return new RGWPSCreateNotif_ObjStore_S3();
+  }
+  RGWOp *op_delete() override {
+    return new RGWPSDeleteNotif_ObjStore_S3();
+  }
+public:
+  explicit RGWHandler_REST_PSNotifs_S3(const rgw::auth::StrategyRegistry& auth_registry) : RGWHandler_REST_S3(auth_registry) {}
+  virtual ~RGWHandler_REST_PSNotifs_S3() = default;
+};
+
+// factory for S3 compliant PubSub REST handlers 
+RGWHandler_REST* RGWRESTMgr_PubSub_S3::get_handler(struct req_state* const s,
+                                                     const rgw::auth::StrategyRegistry& auth_registry,
+                                                     const std::string& frontend_prefix) {
+  if (RGWHandler_REST_S3::init_from_header(s, RGW_FORMAT_XML, true) < 0) {
+    return nullptr;
+  }
+  
+  RGWHandler_REST *handler = nullptr;
+
+  // s3 compliant PubSub API: <bucket name>?notification
+  if (s->info.args.exists("notification")) {
+      handler = new RGWHandler_REST_PSNotifs_S3(auth_registry);
+  } else if (next) {
+      handler = next->get_handler(s, auth_registry, frontend_prefix);
```

Comment:
consider resetting `s->init_state` to its default value here so it's safe to call `init_from_header()` again?